### PR TITLE
Fix improve clipboard arg validation

### LINKED ISSUE #3161: The default title of MsgBox is not correct
* .NET Core Version: 
5.0.100-preview.3.20216.6

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**

The default title is "Microsoft.VisualBasic.Core" instead of the name of the calling assembly.
image

**Expected behavior:**
The default title should be the name of the calling assembly.
image

**Minimal repro:**
Run the following code in a VB winforms project:
```vbnet
MsgBox("Content")
```

### LINKED ISSUE #3163: `Clipboard` API must check for invalid `format`
* .NET Core Version: current master

* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**

Observed in https://dev.azure.com/dnceng/public/_build/results?buildId=622784&view=results

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Windows.Forms.DataObject.OleConverter.GetDataFromOleOther(String format) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\DataObject.cs:line 1435
   at System.Windows.Forms.DataObject.OleConverter.GetDataFromBoundOleDataObject(String format, Boolean& done) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\DataObject.cs:line 1502
   at System.Windows.Forms.DataObject.OleConverter.GetDataFromBoundOleDataObject(String format, Boolean& done) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\DataObject.cs:line 1502
   at System.Windows.Forms.DataObject.OleConverter.GetData(String format, Boolean autoConvert) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\DataObject.cs:line 1677
   at System.Windows.Forms.DataObject.GetData(String format, Boolean autoConvert) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\DataObject.cs:line 174
   at System.Windows.Forms.DataObject.GetData(String format) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\DataObject.cs:line 184
   at System.Windows.Forms.Clipboard.GetData(String format) in F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\Clipboard.cs:line 322
   at System.Windows.Forms.Tests.ClipboardTests.Clipboard_GetData_InvokeMultipleTimes_Success(String format) in F:\workspace\_work\1\s\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\ClipboardTests.cs:line 108
```

Related to #3115.

The following API allow to pass `null`, `string.Empty` or whitespace that all are invalid data formats:
https://github.com/dotnet/winforms/blob/8aeaa419c478dad316edc7ba4103c8abc2bba40f/src/System.Windows.Forms/src/System/Windows/Forms/Clipboard.cs#L317-L326
https://github.com/dotnet/winforms/blob/8aeaa419c478dad316edc7ba4103c8abc2bba40f/src/System.Windows.Forms/src/System/Windows/Forms/Clipboard.cs#L317-L326

**Expected behavior:**

These API must not result in NRE and gracefully handle invalid input. We could throw but this will be breaking.
Instead we should `return null`/`return` respectively.

### PR DESCRIPTION
**Pull request by RussKie**
Resolves #3163

## Proposed changes

- Validate `format` arg and return early, if format invalid
- Split DataObject object definitions, no functional changes

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3167)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Clipboard.cs [hunk 1]
```diff
@@ -397,6 +402,16 @@ public static void SetAudio(Stream audioStream)
...
         {
+            if (string.IsNullOrWhiteSpace(format))
+            {
+                if (format == null)
+                {
+                    throw new ArgumentNullException(nameof(format));
+                }
+
+                throw new ArgumentException(SR.DataObjectWhitespaceEmptyFormatNotAllowed, nameof(format));
+            }
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Clipboard.cs:413]: One of these exceptions would be thrown by `dataObject.SetData(format, data);` call, so just hoisting up the failure exit

**Comment by weltkante** [comment]
@RussKie I think you linked the wrong issue in the OP

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs [hunk 2]
```diff
@@ -231,13 +231,19 @@ public void Clipboard_SetData_Invoke_GetReturnsExpected(string format, object da
...
         [StaTheory]
-        [InlineData(null)]
         [InlineData("")]
+        [InlineData(" ")]
         [InlineData("\t")]
-        public void Clipboard_SetData_NullOrEmptyFormat_DoesNotThrow(string format)
+        public void Clipboard_SetData_EmptyOrWhitespaceFormat_ThrowsArgumentException(string format)
         {
-            Clipboard.SetData(format, null);
+            Assert.Throws<ArgumentException>(() => Clipboard.SetData(format, null));
```
Comment (hughbe) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs:246]: nit: add parameter name here and above

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs [hunk 3]
```diff
@@ -231,13 +231,19 @@ public void Clipboard_SetData_Invoke_GetReturnsExpected(string format, object da
             Assert.True(Clipboard.ContainsData(format));
         }
 
+        [StaFact]
+        public void Clipboard_SetData_NullFormat_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() => Clipboard.SetData(null, null));
+        }
+
         [StaTheory]
```
Comment (hughbe) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs:240]: any reason we prefer StaFact over StaTheory

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs [hunk 4]
```diff
@@ -231,13 +231,19 @@ public void Clipboard_SetData_Invoke_GetReturnsExpected(string format, object da
             Assert.True(Clipboard.ContainsData(format));
         }
 
+        [StaFact]
+        public void Clipboard_SetData_NullFormat_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() => Clipboard.SetData(null, null));
```
Comment (hughbe) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs:237]: nit: can we make `data` non-null just so its clear that the ANE is from `format`, not any other param

**Comment by hughbe** [comment]
I may be missing something, but could you explain how this resolves #3161 which is about the default title of `MsgBox`. Do you mean #3163

**Comment by RussKie** [comment]
> I think you linked the wrong issue in the OP
> Do you mean #3163

Yes, thank you - typo

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs (hunk 3 above)
Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs:240]: Fact -> one test, theory - same test with different values for input parameters
`format == null` results in `ArgumentNullException` (fact), whilst the set of strings result in `ArgumentException` (theory)

There could be another way to convert it to a theory and pass type of expected exception along, but feels like more work with little gain

Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ClipboardTests.cs:240]: Derp. I meant stafact over winformsfact

**Pull request merged**